### PR TITLE
[c++ grpc] Add Bond async client

### PR DESCRIPTION
**Pull request by lalo**
- update test
- helloworld_proxy.h will be codegened

comments:
- ignore Greeter2 in helloworld_proxy.h, temp to get around from some collision with the current codegen
- still have some open questions on what to do with the operators in the proxy class
- I have a redundant use of cq, as part of proxy class and as part of each request that is made see AsyncSayHello, I'm still working on how to simplify that.

### CODE REVIEW

File: cpp/inc/bond/ext/grpc/detail/client_call_data.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#include <thread>
+
+namespace bond { namespace ext { namespace gRPC { namespace detail {
+
+/// @brief Implementation class that hold the state associated with
+/// outgoing incomming calls.
+template <typename TRequest, typename TResponse>
+struct client_unary_call_data : cq_poller_tag
+{
+    typedef std::function<void(TResponse)> CallbackType;
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/client_call_data.h]: We need to pass the returned status to the callback too.

File: cpp/inc/bond/ext/grpc/detail/client_call_data.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    client_unary_call_data(
+        CallbackType cb)
+        : _cb(cb),
+        _response(),
+        _status()
+    {
+        BOOST_ASSERT(cb);
+    }
+
+    void dispatch(grpc::ChannelInterface* channel, ::grpc::CompletionQueue* cq, grpc::RpcMethod method, grpc::ClientContext* context, TRequest request)
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/client_call_data.h]: Minor: lines are long. Can we wrap them?

File: cpp/inc/bond/ext/grpc/detail/client_call_data.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        auto responseReader = new ::grpc::ClientAsyncResponseReader< ::bond::comm::message< ::helloworld::HelloReply>>(channel, cq, method, context, request);
+        responseReader->Finish(&_response, &_status, (void*)this);
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/client_call_data.h]: When does `responseReader` get deleted?

File: cpp/inc/bond/ext/grpc/detail/client_call_data.h (hunk 2 above)
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/client_call_data.h]: The type of request should be `const TRequest&`, and ideally we wouldn't need to copy request ever. We just keep passing references down until we get to `SerializationTraits::Serialize`.

File: cpp/inc/bond/ext/grpc/proxy.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        ~proxy()
+        {
+            Shutdown();
+            Wait();
+        }
+
+        /*
+        proxy(const proxy&) = delete;
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/proxy.h]: These look like they shouldn't be commented out. They signal to the consuming programmer what operations we anticipated on this class.

File: cpp/inc/bond/ext/grpc/proxy.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#include <thread>
+
+namespace bond { namespace ext { namespace gRPC {
+
+    /// @brief Models a gRPC proxy powered by Bond services.
+    ///
+    class proxy
+    {
+    public:
+        proxy(std::unique_ptr<grpc::CompletionQueue> cq)
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/proxy.h]: Hmmm... It should be possible for a bunch of proxies for different services even across different channels to share the same completion queue. A proxy should take a pointer/reference to a cq_poller. We may need to rejigger the abstractions to make this possible.

File: cpp/inc/bond/ext/grpc/proxy.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            _cqPoller.wait();
+        }
+
+        void start()
+        {
+            _cqPoller.start();
+        }
+
+    private:
+
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/proxy.h]: Minor: odd blank line

File: examples/cpp/grpc/helloworld/helloworld.cpp [hunk 7]
```diff
@@ -88,17 +74,28 @@ int main()
...
+
+    bool waitResult = print_event.wait(std::chrono::seconds(10));
+
+    if (waitResult)
         return 0;
-    }
     else
-    {
-        std::cout << "Did not receive correct response. received: " << reply;
         return 1;
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld.cpp]: We should print here that we timed out waiting for the correct response.

**Review by chwarr — CHANGES_REQUESTED** [review]
The biggest thing I see is that there's no way to share a completion queue/cq_poller across multiple proxies.

File: cpp/inc/bond/ext/grpc/proxy.h [hunk 8]
```diff
@@ -14,12 +14,7 @@
     #pragma warning (pop)
 #endif
 
-#include <bond/ext/grpc/detail/cq_poller.h>
-
-#include <boost/assert.hpp>
-#include <boost/optional/optional.hpp>
-#include <memory>
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/proxy.h]: Minor: We still need `#include <memory>`

File: examples/cpp/grpc/helloworld/helloworld.cpp [hunk 9]
```diff
@@ -50,18 +51,24 @@ class GreeterServiceImpl final : public GreeterServiceAsync {
     }
 };
 
-void printAndSet(event* print_event, ::bond::comm::message< ::helloworld::HelloReply> response) {
-    std::string message = response.value().Deserialize().message;
-
-    if (message.compare(std::string("hello world")) == 0)
+void printAndSet(event* print_event, bool* isCorrectResponse, const bond::comm::message< HelloReply>& response, const Status& status) {
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld.cpp]: Minor: line is long. Can we wrap?

File: examples/cpp/grpc/helloworld/helloworld.cpp [hunk 10]
```diff
@@ -50,18 +51,24 @@ class GreeterServiceImpl final : public GreeterServiceAsync {
...
+    if(status.ok())
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld.cpp]: It looks like we don't set `*isCorrectResponse` when `!status.ok()`.

File: examples/cpp/grpc/helloworld/helloworld.cpp [hunk 11]
```diff
@@ -50,18 +51,24 @@ class GreeterServiceImpl final : public GreeterServiceAsync {
...
     {
-        std::cout << "Correct response: " << message;
-        print_event->set();
-    }
-    else
-    {
-        std::cout << "Wrong response";
+        std::string message = response.value().Deserialize().message;
+
+        if (message.compare(std::string("hello world")) == 0)
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld.cpp]: Minor: can simplify to `message.compare("hello world")`, IIRC.

File: examples/cpp/grpc/helloworld/helloworld.cpp [hunk 12]
```diff
@@ -75,9 +82,9 @@ int main()
     std::unique_ptr<bond::ext::gRPC::server> server(builder.BuildAndStart());
 
     std::unique_ptr<grpc::CompletionQueue> cq_(new grpc::CompletionQueue());
-    grpc::CompletionQueue* cq_ptr = cq_.get();
+    std::shared_ptr<io_manager> ioManager(new io_manager(std::move(cq_)));
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld.cpp]: Minor: `auto ioManaged = std::make_shared<io_manager>(std::move(cq_));`, or, if possible: `auto ioManaged = std::make_shared<io_manager>(new grpc::CompletionQueue);`

File: examples/cpp/grpc/helloworld/helloworld.cpp (hunk 7 above)
Comment (chwarr) [reply examples/cpp/grpc/helloworld/helloworld.cpp]: ✔️

File: examples/cpp/grpc/helloworld/helloworld.cpp [hunk 13]
```diff
@@ -87,14 +94,22 @@ int main()
...
+            printAndSet(&print_event, &isCorrectResponse, response, status);
+        };
 
-    greeter.AsyncSayHello(&context, req, cq_ptr, f_print);
+    greeter.AsyncSayHello(&context, req, f_print);
 
     bool waitResult = print_event.wait(std::chrono::seconds(10));
 
-    if (waitResult)
+    if (!waitResult)
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld.cpp]: Minor: can we surround these block with curly braces?

File: cpp/inc/bond/ext/grpc/proxy.h (hunk 4 above)
Comment (chwarr) [reply cpp/inc/bond/ext/grpc/proxy.h]: ✔️

File: cpp/inc/bond/ext/grpc/proxy.h [hunk 14]
```diff
@@ -39,18 +34,16 @@ namespace bond { namespace ext { namespace gRPC {
...
 
         proxy(proxy&&) = default;
         proxy& operator=(proxy&&) = default;
-        */
 
         /// Shutdown the proxy, waiting for all rpc processing to finish.
         void Shutdown()
         {
-            _cqPoller.shutdown();
+            _ioManager->shutdown();
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/proxy.h]: This proxy doesn't exclusively own the `io_manager`, so it shouldn't shut it down. We'll need to figure out some other way to get the behavior we want (the ability to synchronously wait for the outstanding work initiated by this proxy to drain).

File: cpp/inc/bond/ext/grpc/detail/client_call_data.h [hunk 15]
```diff
@@ -49,10 +50,15 @@ struct client_unary_call_data : cq_poller_tag
         BOOST_ASSERT(cb);
     }
 
-    void dispatch(grpc::ChannelInterface* channel, ::grpc::CompletionQueue* cq, grpc::RpcMethod method, grpc::ClientContext* context, TRequest request)
+    void dispatch(grpc::ChannelInterface* channel
+            , io_manager* ioManager
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/client_call_data.h]: Minor: we don't wrap like this in the rest of the Bond code (Comm excepted). Can we switch to

```
void dispatch(grpc::ChannelInterface* channel,
    io_manager* ioManager,
    grpc::RpcMethod method,
...
```

**Review by chwarr — CHANGES_REQUESTED** [review]
The proxy shutting down the `io_manager` that it only has shared ownership of needs to be addressed.

File: cpp/inc/bond/ext/grpc/proxy.h (hunk 14 above)
Comment (lalo) [reply cpp/inc/bond/ext/grpc/proxy.h]: plus I'm not enforcing shutdown also from the dispatch() method

Comment (chwarr) [reply cpp/inc/bond/ext/grpc/proxy.h]: For the initial version of the gRPC++ extensions, let's just not have explicit per-proxy shutdown. If you want to fail all the inflight client calls, shutdown the `io_manager`. If you need to be able to do this per-proxy, create a unique `io_manager` for each proxy (with a limited thread count, likely). We can make this better later, after getting some feedback.

File: cpp/inc/bond/ext/grpc/proxy.h [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+#endif
+
+#include <bond/ext/grpc/io_manager.h>
+#include <memory>
+
+namespace bond { namespace ext { namespace gRPC {
+
+    /// @brief Models a gRPC proxy powered by Bond services.
+    ///
+    class proxy
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/proxy.h]: You're right that this proxy is so thin that it doesn't really need to exist. (There are no sensible operations that one can perform against just a proxy.)

Options I see:

* move the interesting code from the generated `AsyncFoo` methods into a protected method (see below)
* get rid of the proxy class and just generate an `io_manager` member.

Possible protected member

```
// would be inside the class definition, but I'm using extern syntax for this example

template <typename TRequest, typename TResponse, typename Callback>
proxy::send_async(
    ::grpc::RpcMethod& method,
    ::grpc::ClientContext* context,
    const TRequest& request,
    Callback&& cb)
{
    typedef ::bond::ext::gRPC::detail::client_unary_call_data<::bond::comm::message<TRequest>, ::bond::comm::message<TResponse>> CallData;
    auto callData = new CallData(std::forward(cb));
    callData->dispatch(_channel.get(), _ioManager.get(), method, context, request);
}
```

**Review by chwarr — CHANGES_REQUESTED** [review]

File: cpp/inc/bond/ext/grpc/proxy.h (hunk 16 above)
Comment (lalo) [reply cpp/inc/bond/ext/grpc/proxy.h]: I don't see a clear benefit in keeping the proxy, plus we've never had a proxy that is not codegen. I'll add the io_manager.

File: examples/cpp/grpc/helloworld/helloworld_proxy.h [hunk 17]
```diff
@@ -48,13 +54,14 @@ static const char* Greeter_method_names[] = {
 
 Greeter2::GreeterClient::GreeterClient(const std::shared_ptr< ::grpc::ChannelInterface>& channel, std::shared_ptr< ::bond::ext::gRPC::io_manager> ioManager)
     : channel_(channel)
-    , bond::ext::gRPC::proxy(ioManager)
     , rpcmethod_SayHello_(Greeter_method_names[0], ::grpc::RpcMethod::NORMAL_RPC, channel)
-  {}
+  {
+      ioManager_ = ioManager;
```
Comment (chwarr) [inline examples/cpp/grpc/helloworld/helloworld_proxy.h]: Minor: use the initializer list to set this. Also, remember to initialize things in the initializer list in the same order that they're declared.

**Review by chwarr — APPROVED** [review]

**Pull request merged**